```
Error running mint automation
+ mkdir -p 15052-02b11d3/gopath/src/github.com/minio
+ git -C 15052-02b11d3/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 15052-02b11d3/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 15052-02b11d3/gopath/src/github.com/minio/minio fetch --quiet minio pull/15052/head:pr15052
+ git -C 15052-02b11d3/gopath/src/github.com/minio/minio checkout --quiet pr15052
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/15052-02b11d3/gopath
+ make -C 15052-02b11d3/gopath/src/github.com/minio/minio --quiet
go: downloading github.com/minio/madmin-go v1.3.14
go: downloading github.com/minio/cli v1.22.0
go: downloading github.com/json-iterator/go v1.1.12
go: downloading github.com/minio/pkg v1.1.24
go: downloading github.com/dustin/go-humanize v1.0.0
go: downloading github.com/minio/sha256-simd v1.0.0
go: downloading github.com/rs/dnscache v0.0.0-20211102005908-e0241e321417
go: downloading golang.org/x/sys v0.0.0-20220502124256-b6088ccd6cba
go: downloading github.com/minio/sio v0.3.0
go: downloading github.com/fatih/color v1.13.0
go: downloading github.com/secure-io/sio-go v0.3.1
go: downloading github.com/minio/minio-go/v7 v7.0.27
go: downloading github.com/minio/highwayhash v1.0.2
go: downloading golang.org/x/crypto v0.0.0-20220411220226-7b82a4e95df4
go: downloading github.com/Shopify/sarama v1.30.0
go: downloading github.com/golang-jwt/jwt/v4 v4.4.1
go: downloading github.com/xdg/scram v0.0.0-20180814205039-7eeb5667e42c
go: downloading github.com/klauspost/compress v1.15.1
go: downloading github.com/minio/kes v0.19.2
go: downloading github.com/buger/jsonparser v1.1.1
go: downloading github.com/bcicen/jstream v1.0.1
go: downloading github.com/djherbis/atime v1.0.0
go: downloading github.com/felixge/fgprof v0.9.2
go: downloading github.com/google/uuid v1.3.0
go: downloading cloud.google.com/go/storage v1.10.0
go: downloading github.com/inconshreveable/mousetrap v1.0.0
go: downloading github.com/cheggaaa/pb v1.0.29
go: downloading github.com/coreos/go-oidc v2.1.0+incompatible
go: downloading github.com/cespare/xxhash/v2 v2.1.2
go: downloading github.com/klauspost/readahead v1.4.0
go: downloading github.com/Azure/azure-storage-blob-go v0.10.0
go: downloading github.com/hashicorp/golang-lru v0.5.4
go: downloading github.com/dchest/siphash v1.2.1
go: downloading github.com/klauspost/pgzip v1.2.5
go: downloading github.com/go-openapi/loads v0.21.1
go: downloading github.com/minio/dperf v0.4.0
go: downloading github.com/cosnicolaou/pbzip2 v1.0.1
go: downloading github.com/bits-and-blooms/bloom/v3 v3.0.1
go: downloading github.com/klauspost/reedsolomon v1.9.15
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/minio/console v0.18.0
go: downloading github.com/miekg/dns v1.1.48
go: downloading github.com/minio/selfupdate v0.4.0
go: downloading github.com/minio/zipindex v0.2.1
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/philhofer/fwd v1.1.2-0.20210722190033-5c56ac6d0bb9
go: downloading github.com/pierrec/lz4 v2.6.1+incompatible
go: downloading github.com/prometheus/client_golang v1.12.1
go: downloading github.com/prometheus/client_model v0.2.0
go: downloading github.com/prometheus/procfs v0.7.3
go: downloading github.com/rs/cors v1.7.0
go: downloading github.com/shirou/gopsutil/v3 v3.22.2
go: downloading github.com/tinylib/msgp v1.1.7-0.20211026165309-e818a1881b0e
go: downloading github.com/valyala/bytebufferpool v1.0.0
go: downloading github.com/yargevad/filepathx v1.0.0
go: downloading github.com/zeebo/xxh3 v1.0.0
go: downloading go.etcd.io/etcd/api/v3 v3.5.4
go: downloading go.etcd.io/etcd/client/v3 v3.5.4
go: downloading go.uber.org/atomic v1.9.0
go: downloading golang.org/x/oauth2 v0.0.0-20220411215720-9780585627b5
go: downloading google.golang.org/api v0.78.0
go: downloading github.com/coredns/coredns v1.9.2
go: downloading go.uber.org/zap v1.21.0
go: downloading golang.org/x/time v0.0.0-20220224211638-0e9765cccd65
go: downloading github.com/beevik/ntp v0.3.0
go: downloading github.com/go-ldap/ldap/v3 v3.4.3
go: downloading github.com/montanaflynn/stats v0.6.6
go: downloading github.com/minio/simdjson-go v0.4.2
go: downloading github.com/ncw/directio v1.0.5
go: downloading github.com/eclipse/paho.mqtt.golang v1.3.5
go: downloading github.com/elastic/go-elasticsearch/v7 v7.17.1
go: downloading github.com/go-sql-driver/mysql v1.6.0
go: downloading github.com/gomodule/redigo v1.8.8
go: downloading github.com/lib/pq v1.10.4
go: downloading github.com/nats-io/nats.go v1.13.1-0.20220308171302-2f2f6968e98d
go: downloading github.com/nats-io/stan.go v0.10.2
go: downloading github.com/nsqio/go-nsq v1.0.8
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/streadway/amqp v1.0.0
go: downloading github.com/mattn/go-colorable v0.1.12
go: downloading github.com/mattn/go-isatty v0.0.14
go: downloading golang.org/x/sync v0.0.0-20210220032951-036812b2e83c
go: downloading github.com/mattn/go-runewidth v0.0.13
go: downloading github.com/golang/protobuf v1.5.2
go: downloading github.com/minio/csvparser v1.0.0
go: downloading github.com/fraugster/parquet-go v0.10.0
go: downloading github.com/pquerna/cachecontrol v0.1.0
go: downloading gopkg.in/square/go-jose.v2 v2.5.1
go: downloading github.com/alecthomas/participle v0.2.1
go: downloading github.com/xdg/stringprep v1.0.0
go: downloading github.com/klauspost/cpuid/v2 v2.0.11
go: downloading golang.org/x/net v0.0.0-20220425223048-2871e0cb64e4
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading github.com/Azure/go-ntlmssp v0.0.0-20211209120228-48547f28849e
go: downloading github.com/go-asn1-ber/asn1-ber v1.5.4
go: downloading github.com/rivo/uniseg v0.2.0
go: downloading github.com/golang/snappy v0.0.4
go: downloading github.com/bits-and-blooms/bitset v1.2.0
go: downloading github.com/nats-io/nuid v1.0.1
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/modern-go/reflect2 v1.0.2
go: downloading go.etcd.io/etcd/client/pkg/v3 v3.5.4
go: downloading google.golang.org/grpc v1.46.0
go: downloading github.com/go-openapi/analysis v0.21.2
go: downloading github.com/go-openapi/spec v0.20.5
go: downloading github.com/go-openapi/swag v0.21.1
go: downloading google.golang.org/genproto v0.0.0-20220429170224-98d788798c3e
go: downloading github.com/gorilla/websocket v1.5.0
go: downloading github.com/coreos/go-semver v0.3.0
go: downloading golang.org/x/text v0.3.7
go: downloading github.com/Azure/azure-pipeline-go v0.2.2
go: downloading github.com/lestrrat-go/jwx v1.2.19
go: downloading github.com/fatih/structs v1.1.0
go: downloading github.com/rjeczalik/notify v0.9.2
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/prometheus/common v0.34.0
go: downloading github.com/beorn7/perks v1.0.1
go: downloading google.golang.org/protobuf v1.28.0
go: downloading go.uber.org/multierr v1.8.0
go: downloading github.com/rs/xid v1.4.0
go: downloading gopkg.in/ini.v1 v1.66.4
go: downloading github.com/minio/md5-simd v1.1.2
go: downloading github.com/tklauser/go-sysconf v0.3.10
go: downloading cloud.google.com/go v0.100.2
go: downloading github.com/mattn/go-ieproxy v0.0.1
go: downloading github.com/nats-io/nkeys v0.3.0
go: downloading github.com/mailru/easyjson v0.7.7
go: downloading github.com/coreos/go-systemd/v22 v22.3.2
go: downloading github.com/google/pprof v0.0.0-20211214055906-6f57359322fd
go: downloading github.com/apache/thrift v0.15.0
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading github.com/jcmturner/gokrb5/v8 v8.4.2
go: downloading github.com/eapache/go-resiliency v1.2.0
go: downloading github.com/jcmturner/gofork v1.0.0
go: downloading github.com/eapache/queue v1.1.0
go: downloading github.com/rcrowley/go-metrics v0.0.0-20201227073835-cf1acfcdf475
go: downloading github.com/tklauser/numcpus v0.4.0
go: downloading github.com/josharian/intern v1.0.0
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.2-0.20181231171920-c182affec369
go: downloading github.com/go-openapi/strfmt v0.21.2
go: downloading github.com/go-openapi/jsonpointer v0.19.5
go: downloading github.com/lestrrat-go/option v1.0.0
go: downloading github.com/lestrrat-go/iter v1.0.1
go: downloading github.com/lestrrat-go/backoff/v2 v2.0.8
go: downloading github.com/lestrrat-go/blackmagic v1.0.0
go: downloading github.com/lestrrat-go/httpcc v1.0.0
go: downloading github.com/asaskevich/govalidator v0.0.0-20210307081110-f21760c49a8d
go: downloading github.com/mitchellh/mapstructure v1.4.3
go: downloading go.mongodb.org/mongo-driver v1.8.4
go: downloading github.com/go-openapi/errors v0.20.2
go: downloading github.com/oklog/ulid v1.3.1
go: downloading github.com/hashicorp/go-uuid v1.0.2
go: downloading github.com/jcmturner/dnsutils/v2 v2.0.0
go: downloading github.com/jcmturner/aescts/v2 v2.0.0
go: downloading github.com/jcmturner/rpc/v2 v2.0.3
go: downloading github.com/go-openapi/jsonreference v0.20.0
go: downloading github.com/googleapis/gax-go/v2 v2.3.0
go: downloading cloud.google.com/go/iam v0.2.0
go: downloading golang.org/x/xerrors v0.0.0-20220411194840-2f41105eb62f
go: downloading go.opencensus.io v0.23.0
go: downloading github.com/golang/groupcache v0.0.0-20210331224755-41bb18bfe9da
go: downloading github.com/go-stack/stack v1.8.1
go: downloading github.com/go-openapi/runtime v0.23.3
go: downloading github.com/go-openapi/validate v0.21.0
go: downloading github.com/jessevdk/go-flags v1.5.0
go: downloading github.com/minio/mc v0.0.0-20220512134321-aa60a8db1e4d
go: downloading github.com/unrolled/secure v1.10.0
go: downloading github.com/tidwall/gjson v1.14.0
go: downloading github.com/tidwall/pretty v1.2.0
go: downloading github.com/tidwall/match v1.1.1
go: downloading github.com/docker/go-units v0.4.0
go: downloading github.com/gdamore/tcell/v2 v2.4.1-0.20210905002822-f057f0a857a1
go: downloading github.com/pkg/xattr v0.4.5
go: downloading github.com/prometheus/prom2json v1.3.1
go: downloading github.com/olekukonko/tablewriter v0.0.5
go: downloading github.com/posener/complete v1.2.3
go: downloading github.com/briandowns/spinner v1.18.1
go: downloading github.com/charmbracelet/bubbles v0.10.3
go: downloading github.com/rivo/tview v0.0.0-20220216162559-96063d6082f3
go: downloading github.com/charmbracelet/lipgloss v0.5.0
go: downloading golang.org/x/term v0.0.0-20210927222741-03fcf44c2211
go: downloading gopkg.in/h2non/filetype.v1 v1.0.5
go: downloading github.com/minio/colorjson v1.0.2
go: downloading github.com/minio/filepath v1.0.0
go: downloading github.com/google/shlex v0.0.0-20191202100458-e7afc7fbc510
go: downloading maze.io/x/duration v0.0.0-20160924141736-faac084b6075
go: downloading github.com/charmbracelet/bubbletea v0.20.0
go: downloading github.com/navidys/tvxwidgets v0.1.0
go: downloading github.com/muesli/termenv v0.11.1-0.20220212125758-44cd13922739
go: downloading github.com/muesli/reflow v0.3.0
go: downloading github.com/lucasb-eyer/go-colorful v1.2.0
go: downloading github.com/hashicorp/go-multierror v1.1.1
go: downloading github.com/muesli/ansi v0.0.0-20211031195517-c9f0611b6c70
go: downloading github.com/containerd/console v1.0.3
go: downloading github.com/hashicorp/errwrap v1.1.0
go: downloading github.com/gdamore/encoding v1.0.0
go: downloading cloud.google.com/go/compute v1.6.1
# github.com/minio/minio/cmd
cmd/http-stats.go:179:13: serverStats.TotalS3GenericErrors undefined (type ServerHTTPStats has no field or method TotalS3GenericErrors)
cmd/http-stats.go:182:13: serverStats.TotalS3InternalErrors undefined (type ServerHTTPStats has no field or method TotalS3InternalErrors)
cmd/metrics-v2.go:1502:17: httpStats.TotalS3GenericErrors undefined (type ServerHTTPStats has no field or method TotalS3GenericErrors)
cmd/metrics-v2.go:1503:17: httpStats.TotalS3InternalErrors undefined (type ServerHTTPStats has no field or method TotalS3InternalErrors)
cmd/metrics-v2.go:1543:36: httpStats.TotalS3GenericErrors undefined (type ServerHTTPStats has no field or method TotalS3GenericErrors)
cmd/metrics-v2.go:1550:36: httpStats.TotalS3InternalErrors undefined (type ServerHTTPStats has no field or method TotalS3InternalErrors)
cmd/metrics.go:369:35: httpStats.TotalS3GenericErrors undefined (type ServerHTTPStats has no field or method TotalS3GenericErrors)
make: *** [build] Error 2

```

--- a/cmd/admin-handlers.go
+++ b/cmd/admin-handlers.go
@@ -291,6 +291,8 @@ type ServerHTTPStats struct {
 	CurrentS3Requests      ServerHTTPAPIStats `json:"currentS3Requests"`
 	TotalS3Requests        ServerHTTPAPIStats `json:"totalS3Requests"`
 	TotalS3Errors          ServerHTTPAPIStats `json:"totalS3Errors"`
+	TotalS35xxErrors       ServerHTTPAPIStats `json:"totalS35xxErrors"`
+	TotalS34xxErrors       ServerHTTPAPIStats `json:"totalS34xxErrors"`
 	TotalS3Canceled        ServerHTTPAPIStats `json:"totalS3Canceled"`
 	TotalS3RejectedAuth    uint64             `json:"totalS3RejectedAuth"`
 	TotalS3RejectedTime    uint64             `json:"totalS3RejectedTime"`

--- a/cmd/generic-handlers.go
+++ b/cmd/generic-handlers.go
@@ -115,9 +115,11 @@ func setRequestLimitHandler(h http.Handler) http.Handler {
 
 // Reserved bucket.
 const (
-	minioReservedBucket     = "minio"
-	minioReservedBucketPath = SlashSeparator + minioReservedBucket
-	loginPathPrefix         = SlashSeparator + "login"
+	minioReservedBucket              = "minio"
+	minioReservedBucketPath          = SlashSeparator + minioReservedBucket
+	minioReservedBucketPathWithSlash = SlashSeparator + minioReservedBucket + SlashSeparator
+
+	loginPathPrefix = SlashSeparator + "login"
 )
 
 func guessIsBrowserReq(r *http.Request) bool {

--- a/cmd/http-stats.go
+++ b/cmd/http-stats.go
@@ -148,6 +148,8 @@ type HTTPStats struct {
 	currentS3Requests       HTTPAPIStats
 	totalS3Requests         HTTPAPIStats
 	totalS3Errors           HTTPAPIStats
+	totalS34xxErrors        HTTPAPIStats
+	totalS35xxErrors        HTTPAPIStats
 	totalS3Canceled         HTTPAPIStats
 }
 
@@ -178,6 +180,12 @@ func (st *HTTPStats) toServerHTTPStats() ServerHTTPStats {
 	serverStats.TotalS3Errors = ServerHTTPAPIStats{
 		APIStats: st.totalS3Errors.Load(),
 	}
+	serverStats.TotalS34xxErrors = ServerHTTPAPIStats{
+		APIStats: st.totalS34xxErrors.Load(),
+	}
+	serverStats.TotalS35xxErrors = ServerHTTPAPIStats{
+		APIStats: st.totalS35xxErrors.Load(),
+	}
 	serverStats.TotalS3Canceled = ServerHTTPAPIStats{
 		APIStats: st.totalS3Canceled.Load(),
 	}
@@ -186,27 +194,29 @@ func (st *HTTPStats) toServerHTTPStats() ServerHTTPStats {
 
 // Update statistics from http request and response data
 func (st *HTTPStats) updateStats(api string, r *http.Request, w *logger.ResponseWriter) {
-	// A successful request has a 2xx response code or < 4xx response
-	successReq := w.StatusCode >= 200 && w.StatusCode < 400
-
-	if !strings.HasSuffix(r.URL.Path, prometheusMetricsPathLegacy) ||
-		!strings.HasSuffix(r.URL.Path, prometheusMetricsV2ClusterPath) ||
-		!strings.HasSuffix(r.URL.Path, prometheusMetricsV2NodePath) {
-		st.totalS3Requests.Inc(api)
-		if !successReq {
-			switch w.StatusCode {
-			case 0:
-			case 499:
-				// 499 is a good error, shall be counted as canceled.
-				st.totalS3Canceled.Inc(api)
-			default:
-				st.totalS3Errors.Inc(api)
-			}
-		}
+	// Ignore non S3 requests
+	if strings.HasSuffix(r.URL.Path, minioReservedBucketPathWithSlash) {
+		return
 	}
 
 	// Increment the prometheus http request response histogram with appropriate label
 	httpRequestsDuration.With(prometheus.Labels{"api": api}).Observe(w.TimeToFirstByte.Seconds())
+
+	code := w.StatusCode
+
+	switch {
+	case code == 0:
+	case code == 499:
+		// 499 is a good error, shall be counted as canceled.
+		st.totalS3Canceled.Inc(api)
+	case code >= http.StatusBadRequest:
+		st.totalS3Errors.Inc(api)
+		if code >= http.StatusInternalServerError {
+			st.totalS35xxErrors.Inc(api)
+		} else {
+			st.totalS34xxErrors.Inc(api)
+		}
+	}
 }
 
 // Prepare new HTTPStats structure

--- a/cmd/metrics-v2.go
+++ b/cmd/metrics-v2.go
@@ -603,7 +603,27 @@ func getS3RequestsErrorsMD() MetricDescription {
 		Namespace: s3MetricNamespace,
 		Subsystem: requestsSubsystem,
 		Name:      errorsTotal,
-		Help:      "Total number S3 requests with errors",
+		Help:      "Total number S3 requests with (4xx and 5xx) errors",
+		Type:      counterMetric,
+	}
+}
+
+func getS3Requests4xxErrorsMD() MetricDescription {
+	return MetricDescription{
+		Namespace: s3MetricNamespace,
+		Subsystem: requestsSubsystem,
+		Name:      "4xx_" + errorsTotal,
+		Help:      "Total number S3 requests with (4xx) errors",
+		Type:      counterMetric,
+	}
+}
+
+func getS3Requests5xxErrorsMD() MetricDescription {
+	return MetricDescription{
+		Namespace: s3MetricNamespace,
+		Subsystem: requestsSubsystem,
+		Name:      "5xx_" + errorsTotal,
+		Help:      "Total number S3 requests with (5xx) errors",
 		Type:      counterMetric,
 	}
 }
@@ -1488,7 +1508,9 @@ func getHTTPMetrics() *MetricsGroup {
 		metrics = make([]Metric, 0, 3+
 			len(httpStats.CurrentS3Requests.APIStats)+
 			len(httpStats.TotalS3Requests.APIStats)+
-			len(httpStats.TotalS3Errors.APIStats))
+			len(httpStats.TotalS3Errors.APIStats)+
+			len(httpStats.TotalS35xxErrors.APIStats)+
+			len(httpStats.TotalS34xxErrors.APIStats))
 		metrics = append(metrics, Metric{
 			Description: getS3RejectedAuthRequestsTotalMD(),
 			Value:       float64(httpStats.TotalS3RejectedAuth),
@@ -1531,6 +1553,20 @@ func getHTTPMetrics() *MetricsGroup {
 		for api, value := range httpStats.TotalS3Errors.APIStats {
 			metrics = append(metrics, Metric{
 				Description:    getS3RequestsErrorsMD(),
+				Value:          float64(value),
+				VariableLabels: map[string]string{"api": api},
+			})
+		}
+		for api, value := range httpStats.TotalS35xxErrors.APIStats {
+			metrics = append(metrics, Metric{
+				Description:    getS3Requests5xxErrorsMD(),
+				Value:          float64(value),
+				VariableLabels: map[string]string{"api": api},
+			})
+		}
+		for api, value := range httpStats.TotalS34xxErrors.APIStats {
+			metrics = append(metrics, Metric{
+				Description:    getS3Requests4xxErrorsMD(),
 				Value:          float64(value),
 				VariableLabels: map[string]string{"api": api},
 			})

--- a/docs/metrics/prometheus/list.md
+++ b/docs/metrics/prometheus/list.md
@@ -52,7 +52,9 @@ These metrics can be from any MinIO server once per collection.
 | `minio_node_process_uptime_seconds`          | Uptime for MinIO process per node in seconds.                                                                       |
 | `minio_node_syscall_read_total`              | Total read SysCalls to the kernel. /proc/[pid]/io syscr                                                             |
 | `minio_node_syscall_write_total`             | Total write SysCalls to the kernel. /proc/[pid]/io syscw                                                            |
-| `minio_s3_requests_error_total`              | Total number S3 requests with errors                                                                                |
+| `minio_s3_requests_errors_total`             | Total number S3 requests with 4xx and 5xx errors                                                                    |
+| `minio_s3_requests_4xx_errors_total`         | Total number S3 requests with 4xx errors                                                                            |
+| `minio_s3_requests_5xx_errors_total`         | Total number S3 requests with 5xx errors                                                                            |
 | `minio_s3_requests_inflight_total`           | Total number of S3 requests currently in flight                                                                     |
 | `minio_s3_requests_total`                    | Total number S3 requests                                                                                            |
 | `minio_s3_time_ttfb_seconds_distribution`    | Distribution of the time to first byte across API calls.                                                            |

